### PR TITLE
Mesh API

### PR DESCRIPTION
**Pull request by archanasri**
This PR consolidates the following executables:
- ProjectPointsOntoMesh
- ComputeGeodesicDistanceFromVerticesToPoints
- ComputeMeshGeodesics
- GetFeatureVolume
- GradientTesting

Added python bindings. Added MeshTests, shapeworksTests and PythonTests

### CODE REVIEW

File: Libs/Mesh/Mesh.h [hunk 1]
```diff
@@ -87,6 +88,11 @@ class Mesh
   /// computes cell normals and orients them such that they point in the same direction
   Mesh& generateNormals();
 
+  double projectPoint(Point3 point);
```
Comment (cchriste) [inline Libs/Mesh/Mesh.h:91]: /// doc string needed

File: Libs/Mesh/Mesh.cpp [hunk 2]
```diff
@@ -346,23 +347,60 @@ PhysicalRegion Mesh::boundingBox() const
   return bbox;
 }
 
+Mesh& Mesh::fix(bool smoothBefore, bool smoothAfter, double lambda, int iterations, bool decimate, double percentage)
+{
+	FEVTKimport import;
```
Comment (cchriste) [inline Libs/Mesh/Mesh.cpp:352]: tabs instead of spaces?

File: Libs/Mesh/CMakeLists.txt [hunk 3]
```diff
@@ -62,16 +62,6 @@ install(TARGETS Mesh
...
-ADD_EXECUTABLE(ComputeMeshGeodesics ComputeMeshGeodesics.cxx)
-TARGET_LINK_LIBRARIES(ComputeMeshGeodesics ${ITK_LIBRARIES} Mesh tinyxml)
-INSTALL(TARGETS ComputeMeshGeodesics RUNTIME DESTINATION bin)
-
-# Uses custom TriMesh functions without making an instance of meshFIM
-#ADD_EXECUTABLE(GetFeatureVolume GetFeatureVolume.cxx)
-#TARGET_LINK_LIBRARIES(GetFeatureVolume ${ITK_LIBRARIES} ${VXL_LIBRARIES} Mesh tinyxml)
-#INSTALL(TARGETS GetFeatureVolume RUNTIME DESTINATION bin)
-
 #ADD_EXECUTABLE(ComputeGeodesicDistanceToCurve ComputeGeodesicDistanceToCurve.cxx)
```
Comment (cchriste) [inline Libs/Mesh/CMakeLists.txt:65]: still needed or can these commented out executables simply be removed?

File: Libs/Python/ShapeworksPython.cpp [hunk 4]
```diff
@@ -965,6 +965,22 @@ PYBIND11_MODULE(shapeworks_py, m)
...
+
+  .def("projectPoint",
+       [](Mesh &mesh, std::vector<double> p) -> decltype(auto) {
+         return mesh.projectPoint(Point({p[0], p[1], p[2]}));
+       },
+       "point"_a)
+
+  .def("geodesicDistance",
+       &Mesh::geodesicDistance,
+       "computes geodesic distance between two vertices (specified by their indices) on mesh",
```
Comment (cchriste) [inline Libs/Python/ShapeworksPython.cpp:981]: good help line 👍

File: Testing/PythonTests/geodesic.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+import sys
+from shapeworks import *
+
+def geodesicTest():
+  mesh = Mesh(os.environ["DATA"] + "/ellipsoid_0.ply")
+  dist = mesh.geodesicDistance(10, 20)
+
+  return abs(dist - 1.0083) < 1e-4
+
+val = geodesicTest()
```
Comment (cchriste) [inline Testing/PythonTests/geodesic.py:11]: Please use try/except for this and all other python tests.

Example:
```python
try:
  if not testFunction():
    print("testName failed")
    sys.exit(1)
except RuntimeError:
  print("testName failed (exception)")
  sys.exit(1)
```

File: Libs/Mesh/Mesh.h [hunk 6]
```diff
@@ -107,6 +113,12 @@ class Mesh
   /// number of faces
   vtkIdType numFaces() const { return mesh->GetNumberOfCells(); }
 
+  /// returns matrix with vertex data
```
Comment (cchriste) [inline Libs/Mesh/Mesh.h:116]: Can you please describe this more specifically? Even looking at the definition I still don't understand what it's doing.

**Review by cchriste — CHANGES_REQUESTED** [review]
Seems to be missing tests for projectPoint and I think another one of the python functions.
Couple of other minor thing noted below.
But looks good overall.

File: Libs/Mesh/Mesh.h (hunk 1 above)
Comment (archanasri) [reply Libs/Mesh/Mesh.h:91]: Done

File: Libs/Mesh/Mesh.cpp (hunk 2 above)
Comment (archanasri) [reply Libs/Mesh/Mesh.cpp:352]: Done

File: Testing/PythonTests/geodesic.py (hunk 5 above)
Comment (archanasri) [reply Testing/PythonTests/geodesic.py:11]: Done

File: Libs/Mesh/CMakeLists.txt (hunk 3 above)
Comment (archanasri) [reply Libs/Mesh/CMakeLists.txt:65]: Removed.

File: Libs/Mesh/Mesh.h (hunk 6 above)
Comment (archanasri) [reply Libs/Mesh/Mesh.h]: Done

**Review by cchriste — APPROVED** [review]
All tests added and verified for python, mesh, and command line. Looks good to merge.

**Pull request merged**